mypy/build.py

```diff
@@ -891,36 +930,65 @@ def find() -> Optional[str]:
                 if isdir:
                     dirs.append(dir)
             find_module_dir_cache[dir_chain, lib_path] = dirs
-        candidate_base_dirs = find_module_dir_cache[dir_chain, lib_path]
-
+        third_party_dirs = []
+        # Third-party stub/typed packages
+        for pkg_dir in site_packages_dirs:
+            stub_name = components[0] + '-stubs'
+            typed_file = os.path.join(pkg_dir, components[0], 'py.typed')
+            stub_dir = os.path.join(pkg_dir, stub_name)
+            if os.path.isdir(stub_dir):
+                stub_components = [stub_name] + components[1:]
+                path = os.path.join(pkg_dir, *stub_components[:-1])
+                if os.path.isdir(path):
+                    third_party_dirs.append(path)
+            elif os.path.isfile(typed_file):
+                path = os.path.join(pkg_dir, dir_chain)
+                third_party_dirs.append(path)
+
+        candidate_base_dirs = tuple(third_party_dirs + find_module_dir_cache[dir_chain, lib_path])
```

Comment:
Why the tuple conversion?

Reply:
Ah, this is just due to it being in my previous commits. I will get rid of it when I probably end up rebasing this on #4623.

Reply:
This looks pretty close to done to me - looks like your patch got a lot cleaner after adding the refactor and removing it again :). So maybe it will get in before #4623 after all?

Reply:
Not sure. It is probably easier for me to rebase since my implementation is just one commit to rollback and apply, but I would prefer to get this PR in first. :)